src/main/java/fi/hsl/transitdata/tripupdate/processing/TripIdCache.java

```diff
@@ -0,0 +1,121 @@
+package fi.hsl.transitdata.tripupdate.processing;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import fi.hsl.common.transitdata.PubtransFactory;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Google needs trip ID for cancellations.
+ * As Transitdata pipeline does not handle trip ids, we fetch and cache trip ID from Digitransit API to use for cancellations.
+ */
+public class TripIdCache {
+    private static final Logger logger = LoggerFactory.getLogger(TripIdCache.class);
+
+    private static final String FUZZY_TRIP_QUERY = "{ fuzzyTrip(route: \"HSL:%s\", direction: %d, date: \"%s\", time: %d) { gtfsId } }";
+
+    private LoadingCache<TripDetails, Optional<String>> tripIdCache = CacheBuilder.newBuilder().expireAfterAccess(1, TimeUnit.HOURS).build(new CacheLoader<TripDetails, Optional<String>>() {
+        @Override
+        public Optional<String> load(TripDetails tripDetails) throws Exception {
+            return getTripIdFromDigitransitAPI(tripDetails.routeId, tripDetails.operatingDay, tripDetails.startTime, tripDetails.directionId);
+        }
+    });
+
+    public Optional<String> getTripId(String routeId, String operatingDay, String startTime, int directionId) {
+        try {
+            return tripIdCache.get(new TripDetails(routeId, operatingDay, startTime, directionId));
+        } catch (ExecutionException e) {
+            logger.warn("Failed to fetch trip id for {} / {} / {} / {}", routeId, operatingDay, startTime, directionId, e);
+            return Optional.empty();
+        }
+    }
+
+    private static Optional<String> getTripIdFromDigitransitAPI(String routeId, String operatingDay, String startTime, int directionId) throws IOException {
+        //TODO: maybe this URL should be configurable?
+        URLConnection connection = new URL("https://api.digitransit.fi/routing/v1/routers/hsl/index/graphql").openConnection();
+        connection.setDoOutput(true);
+        connection.setRequestProperty("Content-Type", "application/graphql");
+        connection.connect();
+
+        try (OutputStream os = connection.getOutputStream()) {
+            os.write(String.format(FUZZY_TRIP_QUERY, routeId, PubtransFactory.joreDirectionToGtfsDirection(directionId), operatingDay, startTimeToSeconds(startTime)).getBytes(StandardCharsets.UTF_8));
+        }
+
+        try (InputStream is = connection.getInputStream()) {
```

Comment:
catch and log something like "Failed to read input stream from digitransitRouteAPI